include/pybind11/pybind11.h

```diff
@@ -50,13 +50,25 @@ class cpp_function : public function {
         initialize(f, f, extra...);
     }
 
+#if !defined(_MSC_VER)
     /// Construct a cpp_function from a lambda function (possibly with internal state)
     template <typename Func, typename... Extra,
              typename FuncType = typename detail::remove_class<decltype(&std::remove_reference<Func>::type::operator())>::type>
     cpp_function(Func &&f, const Extra&... extra) {
```

Comment:
GCC and clang 4 (or possibly just clang 4 in C++1z mode) also needed a `remove_reference` and `!std::is_function` test to avoid compilation errors, but I think that approach is nicer (and should keep working as MSVC gains C++17 support).